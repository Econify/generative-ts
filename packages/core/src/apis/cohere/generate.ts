/* eslint-disable camelcase */
import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/lib/Either.js";

import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { FnTemplate } from "../../utils/Template";

/**
 * @category Requests
 * @category Cohere Generate
 */
export interface CohereGenerateOptions extends ModelRequestOptions {
  num_generations?: number;
  stream?: boolean;
  max_tokens?: number;
  truncate?: "NONE" | "START" | "END";
  temperature?: number;
  seed?: number;
  preset?: string;
  end_sequences?: string[];
  stop_sequences?: string[];
  k?: number;
  p?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  return_likelihoods?: "GENERATION" | "ALL" | "NONE";
  logit_bias?: { [token_id: number]: number }; // on bedrock but not cohere /generate?
}

/**
 * @category Templates
 * @category Cohere Generate
 */
export const CohereGenerateTemplate = new FnTemplate(
  ({
    prompt,
    num_generations,
    stream,
    max_tokens,
    truncate,
    temperature,
    seed,
    preset,
    end_sequences,
    stop_sequences,
    k,
    p,
    frequency_penalty,
    presence_penalty,
    return_likelihoods,
    logit_bias,
  }: CohereGenerateOptions) => {
    return JSON.stringify(
      {
        prompt,
        num_generations,
        stream,
        max_tokens,
        truncate,
        temperature,
        seed,
        preset,
        end_sequences,
        stop_sequences,
        k,
        p,
        frequency_penalty,
        presence_penalty,
        return_likelihoods,
        logit_bias,
      },
      null,
      2,
    );
  },
);

const CohereGenerateResponseCodec = t.intersection([
  t.type({
    id: t.string,
    prompt: t.string,
    generations: t.array(
      t.type({
        id: t.string,
        text: t.string,
        finish_reason: t.string,
      }),
    ),
  }),
  t.partial({
    meta: t.type({
      api_version: t.type({
        version: t.string,
      }),
      billed_units: t.type({
        input_tokens: t.number,
        output_tokens: t.number,
      }),
    }),
  }),
]);

/**
 * @category Responses
 * @category Cohere Generate
 */
export interface CohereGenerateResponse
  extends TypeOf<typeof CohereGenerateResponseCodec> {}

export function isCohereGenerateResponse(
  response: unknown,
): response is CohereGenerateResponse {
  return !isLeft(CohereGenerateResponseCodec.decode(response));
}

export interface CohereGenerateApi
  extends ModelApi<CohereGenerateOptions, CohereGenerateResponse> {}

/**
 *
 * ## Reference
 * [Cohere Generate](https://docs.cohere.com/reference/generate)
 *
 * ## Providers using this API
 * - {@link createCohereModelProvider | Cohere}
 * - {@link createAwsBedrockModelProvider | AWS Bedrock}
 *
 * @category APIs
 * @category Cohere Generate
 * @category Provider: AWS Bedrock
 * @category Provider: Cohere
 *
 */
export const CohereGenerateApi: CohereGenerateApi = {
  requestTemplate: CohereGenerateTemplate,
  responseGuard: isCohereGenerateResponse,
  name: "COHERE-GENERATE",
};
