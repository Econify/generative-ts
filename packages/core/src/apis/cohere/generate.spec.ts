import { CohereGenerateApi, CohereGenerateOptions } from "./generate";

function render(context: Omit<CohereGenerateOptions, "modelId">) {
  try {
    const rendered = CohereGenerateApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("CohereGenerateApi:", () => {
  /**
   * FewShotRequestOptions (prompt):
   */

  test("$prompt", () => {
    const rendered = render({
      $prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      num_generations: 3,
      max_tokens: 1000,
      truncate: "END",
      temperature: 0.7,
      seed: 1234,
      end_sequences: ["mock-end-seq"],
      stop_sequences: ["mock-stop-seq"],
      k: 5,
      p: 0.9,
      frequency_penalty: 0.5,
      presence_penalty: 0.3,
      return_likelihoods: "ALL",
      logit_bias: { 50256: -100 },
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * Special cases:
   */

  test("$prompt, stop_sequences and end_sequences", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      stop_sequences: ["mock-stop-seq"],
      end_sequences: ["mock-end-seq"],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, truncate as NONE", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      truncate: "NONE",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, return_likelihoods as NONE", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      return_likelihoods: "NONE",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, logit_bias with multiple values", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      logit_bias: { 50256: -100, 50257: 50 },
    });

    expect(rendered).toMatchSnapshot();
  });
});
