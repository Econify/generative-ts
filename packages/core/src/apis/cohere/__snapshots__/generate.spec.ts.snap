// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CohereGenerateApi: all options 1`] = `
{
  "end_sequences": [
    "mock-end-seq",
  ],
  "frequency_penalty": 0.5,
  "k": 5,
  "logit_bias": {
    "50256": -100,
  },
  "max_tokens": 1000,
  "num_generations": 3,
  "p": 0.9,
  "presence_penalty": 0.3,
  "prompt": "mock-prompt",
  "return_likelihoods": "ALL",
  "seed": 1234,
  "stop_sequences": [
    "mock-stop-seq",
  ],
  "temperature": 0.7,
  "truncate": "END",
}
`;

exports[`CohereGenerateApi: prompt 1`] = `
{
  "prompt": "mock-prompt",
}
`;

exports[`CohereGenerateApi: prompt, logit_bias with multiple values 1`] = `
{
  "logit_bias": {
    "50256": -100,
    "50257": 50,
  },
  "prompt": "mock-prompt",
}
`;

exports[`CohereGenerateApi: prompt, return_likelihoods as NONE 1`] = `
{
  "prompt": "mock-prompt",
  "return_likelihoods": "NONE",
}
`;

exports[`CohereGenerateApi: prompt, stop_sequences and end_sequences 1`] = `
{
  "end_sequences": [
    "mock-end-seq",
  ],
  "prompt": "mock-prompt",
  "stop_sequences": [
    "mock-stop-seq",
  ],
}
`;

exports[`CohereGenerateApi: prompt, truncate as NONE 1`] = `
{
  "prompt": "mock-prompt",
  "truncate": "NONE",
}
`;
